### ISSUE TITLE
noisy warning message NB treating all values as 'exists' - no longer sparse

### ISSUE THREAD

**Issue by mckelvin** [manahl/arctic#301]
#### Arctic Version

```
v1.35.0
```

#### Arctic Store

```
TickStore
```

#### Platform and version

Not related

#### Description of problem and/or code sample that reproduces the issue

```python
logging.basicConfig(level=logging.INFO)

# Import tick by tick
for i in loops:
    df = pd.DataFrame(
        {"price": [price]}, index=[datetime.datetime.now(mktz("UTC"))]
    )
    tick_library.write("symbol1", df)
...

# Compress daily ticks by rewriting
date_range = DateRange("2016-12-09 08:00", 2016-12-10 08:00, CLOSED_OPEN)
df_backup = tick_library.read("symbol1", date_range)
tick_library.delete("symbol1", date_range)
tick_library.write("symbol1", df_backup)
```

The way how I use the tick library has been described in https://github.com/manahl/arctic/issues/192 , and I have to compress ticks every few hours before my disk was empty.

During the rewrite step,

```
tick_library.write("symbol1", df_backup)
```

A warning
```
2016-12-09 17:20:30,768 [WARNING] [arctic.tickstore.tickstore] NB treating all values as 'exists' - no longer sparse
```

will raise constantly, which is quite noisy. This warning will raise if the second argument in `tick_library.write` is a `pandas.DataFrame`, even my df is not spare at all . I think it's better to warn just once by using the Python `warning` module.

BTW, is there any better way to compress existed data instead of doing **read-delete-write** manually?

Thanks.

**Comment by bmoscon** [comment]
it is a noisy log, that could/should be removed. 

Not sure what your 2nd question is about. The data is compressed on write in tickstore (see https://github.com/manahl/arctic/blob/master/arctic/tickstore/tickstore.py#L635)

**Comment by mckelvin** [comment]
> it is a noisy log, that could/should be removed.

I wish it would be rename to `.debug` or use the `warning` module.

The second problem is:

Support there're 10000 ticks. It can be write into Arctic by 2 ways:

1. Write tick by tick
2. Write all ticks at once

In the first way, each tick is compress into a bucket, so there're many buckets with just only one tick in mongdb. The compress rate is fairly bad.
In the second way, all ticks are compressed into just one bucket and the compress rate is good.

In my use case I have to convert records (with `one bucket one tick` buckets, which ) to records (with `one bucket multi ticks` buckets) by:

1. Read out old records
2. Delete old records
3. Re-write old records

Is there any better way.

**Comment by bmoscon** [comment]
dont write them all at once? Why cant you build up the dataframe or dict and write it less frequently? I dont use tickstore so maybe @jamesblackburn has a better idea/solution

**Comment by mckelvin** [comment]
@bmoscon Because these are realtime ticks, and should be written to Arctic ASAP. Though I can write arctic every few (say 10 seconds) by building up ticks in 10s into one bucket (one dataframe). The compress rate is still not good.

**Comment by bmoscon** [comment]
I'll let @jamesblackburn or @richardbounds comment on this - not sure what the best way is to get compression on the ticks if you write them as soon as they come in

**Comment by jamesblackburn** [comment]
> Is there any better way.

We use kafka as our real-time tick queue, and batch ~3 hours of ticks to the store at once. The tick flow looks something like this:
http://www.slideshare.net/JamesBlackburn1/building-a-time-series-database/35

> This warning will raise if the second argument in tick_library.write is a pandas.DataFrame, even my df is not spare at all . I think it's better to warn just once by using the Python warning module.

Using the warning module would be reasonable.  We don't use this code-path as the fields in our tick data is really sparse, so we don't want the column data to be fully materialised on write.

**Comment by richardbounds** [comment]
Wow. I wouldn't recommend writing tick-by-tick to Arctic, it really wasn't designed for that at all. The closest we get to that is writing 1-minute OHLC bars to VersionStore by calling store.append() for a few thousand tickers every minute.

As James says, you need to batch the tick data somehow and write to TickStore in batches - we use Kafka, which works well for a relatively high throughput system. A lighter-weight option might be to write tiny 'tick' documents to a capped collection in Mongo and offload them periodically to Arctic.

**Comment by mckelvin** [comment]
Hi @richardbounds , thanks for your advice. I've changed the way how I use arctic:

Now ticks will be:

1. wrote to csv files using the [csv module](https://docs.python.org/2/library/csv.html) in Python, which supports buffered-write. So it won't bring too much disk IO. Ticks in csv files will be loaded and imported to arctic library (tick library) after trading hours.
2. published to a Redis PubSub service (just like what Kafka do). Another worker will subscribe the ticks, compute 1-min OHLC bars and write to arctic bar by bar instantly. Currently I'm also using the **tick library** to store the OHLC bars. At the end of a day, all 1-min OHLC bars in that day will be compressed by the read-delete-write process.

I think the issue has already been solved by changing the way how I use Arctic, so I'm going to close the issue. Thanks you guys.
